Add correct implementation of to_python in ES adapters

AmitPhulera opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
Going down the line we aim to reach a point where we directly pass in the High-level objects to the adapter and it should be able to transform and index it without any issues.
This is one of the steps in that direction.
Ideally, we would want `from_python` to deal take python objects as inputs and then transform them. Since pillows entirely work on dicts, it becomes a requirement for now to support both python objects and dicts in the index function of the adapter. So we introduce a new function `from_multi` which can take in multiple formats and do the transformation on them and return the required dict. `from_python` then calls `from_multi` but `from_python` is only restricted to use high-level python objects.

Review by commit 🐟 

(I might adding more commits that will replace transform functions from tests and code, I might add them as separated PR
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-14329
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The changes should be a no-op as the added functions are copies of transform functions but changes have been tested locally
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Changes covered by tests
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->



<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

AmitPhulera committed:
add new exception classes

AmitPhulera committed:
create from_multi method in user_adapter

AmitPhulera committed:
make from_python use from_multi

AmitPhulera committed:
add tests for to_multi in user_adapter

AmitPhulera committed:
update tests that use transform_user_for_elasticsearch

AmitPhulera committed:
add from_multi in ElasticDocumentAdapter

AmitPhulera committed:
let index use from_multi

AmitPhulera committed:
update user/signals.py to use native index instead of transform function

AmitPhulera committed:
add to_python and to_multi in Domain

AmitPhulera committed:
add tests

AmitPhulera committed:
add to_python and to_multi in apps

AmitPhulera committed:
add tests

AmitPhulera committed:
add to_python and to_multi in cases

AmitPhulera committed:
add tests

AmitPhulera committed:
add to_python and to_multi in case_search

AmitPhulera committed:
add tests

AmitPhulera committed:
add to_python and to_multi in forms

AmitPhulera committed:
add tests

millerdev commented on corehq/apps/es/client.py:
```diff
@@ -442,6 +442,20 @@ def from_python(cls, doc):
         raise NotImplementedError(f"{cls.__name__} is abstract")
 
+    @classmethod
+    def from_multi(cls, doc):
```
What do you think of renaming `from_multi` to `from_dict` and have `from_python` do something like this?
```py
def from_python(cls, doc):
    return cls.from_dict(doc.to_json())
```

`from_dict` should not do any validation. It should assume that the dict being passed in is a valid data for this adapter. It is the responsibility of the caller to do validation if it is necessary in that context.

AmitPhulera replied:
I love it @millerdev. I will update the code accordingly.

>  My original understanding was that the from_multi method would just be an implementation detail on some adapter subclasses (only the ones that need to accept multiple types), and would get called from inside those adapter's from_python method. Thoughts?

I also thought of the same initially but given that all the adapters need this functionality so I moved it up in the parent class like `to_python`. Would love to offline on this and hash out the implementation details.

millerdev commented on corehq/apps/es/domains.py:
```diff
@@ -56,8 +62,48 @@ def mapping(self):
+
+        def _verify_and_return_domain_as_dict():
+            if isinstance(domain, dict):
+                return deepcopy(domain)
```
Suggestion: use `domain.copy()` (shallow copy) instead of `deepcopy(domain)` if possible.

This would require adjusting some code below such as
```py
doc_ret['deployment']['countries'] = []
```

millerdev commented on corehq/apps/es/domains.py:
```diff
@@ -56,8 +62,48 @@ def mapping(self):
+
+        doc_ret = _verify_and_return_domain_as_dict()
+        sub = Subscription.visible_objects.filter(subscriber__domain=doc_ret['name'], is_active=True)
+        doc_ret['deployment'] = doc_ret.get('deployment', None) or {}
```
nit
```suggestion
        doc_ret['deployment'] = doc_ret.get('deployment') or {}
```

joxl commented on corehq/apps/users/signals.py:
```diff
@@ -29,13 +29,7 @@ def _update_user_in_es(couch_user):
-
-
-def _transform_for_es(couch_user):
-    """Implemented separately for test patching."""
```
FYI, [this test](https://github.com/dimagi/commcare-hq/blob/3f4877a8d31/corehq/apps/users/tests/test_signals.py#L134) also needs to be updated (remove the `simple_doc` function and `with patch(...)` block).

joxl commented on corehq/apps/es/tests/test_client.py:
```diff
@@ -1795,6 +1820,354 @@ def test_create_document(self):
+            username='webman', domain=cls.domain, password="***********",
+            created_by=None, created_via=None
+        )
+
+    @classmethod
+    def tearDownClass(cls):
+        delete_all_users()
+        super().tearDownClass()
```
nit (I'm not really sure if this gains anything functionally, but it's a little cleaner)

```suggestion
        cls.addClassCleanup(delete_all_users)
```

AmitPhulera replied:
Yep. Thanks for the suggestion.

joxl commented on corehq/apps/es/tests/test_client.py:
```diff
@@ -1,43 +1,68 @@
 
+from corehq.apps.app_manager.tests.app_factory import AppFactory
+from corehq.apps.domain.shortcuts import create_domain
+from corehq.apps.es.apps import app_adapter
+from corehq.apps.es.case_search import case_search_adapter
+from corehq.apps.es.cases import case_adapter
+from corehq.apps.es.domains import domain_adapter
+from corehq.apps.es.forms import form_adapter
+from corehq.apps.es.users import user_adapter
```
It would be helpful if the tests for these adapters were moved to new test modules. Putting them in `test_client.py` is a bit confusing because the code being tested isn't from `client.py`. Adding new test modules (`test_apps.py`, `test_cases.py`, `test_users.py`, etc) has the added benefit of providing a home for future tests that are currently missing for those modules.

AmitPhulera replied:
I was actually tied between moving them to new files or to keep them in `test_client.py`  and my rationale for putting them in the client file was that we are testing a specific functionally of the client. 
But I see your suggestion makes more sense of having separate files for every adapter.

AmitPhulera commented on corehq/apps/es/users.py:
```diff
@@ -74,8 +81,52 @@ def mapping(self):
+                return user.to_json()
+            raise UnknownDocException(CouchUser, user)
+
+        user_dict = _verify_and_return_user_as_dict()
```
Yeah. I will remove it. This was done earlier and dict validation was not included in other adapters.

AmitPhulera commented on corehq/apps/es/tests/test_client.py:
```diff
@@ -1795,6 +1820,354 @@ def test_create_document(self):
+        self.addCleanup(user_adapter.delete, self.user._id)
+
+        commcare_user = user_adapter.to_json(self.user)
+        es_user = user_adapter.search({})['hits']['hits'][0]['_source']
```
I think it should work. I think I wanted to see that there is only one doc in that index.

AmitPhulera commented on corehq/apps/es/users.py:
```diff
@@ -74,8 +81,52 @@ def mapping(self):
+            if isinstance(user, dict):
+                try:
+                    CouchUser.wrap_correctly(user)
+                    return deepcopy(user)
```
> Can this be replaced with a shallow copy? 

Sure. Sounds like a good optimization.


> In the case of user_dict['user_data_es'].append(...) below, it could do something like

Just for my clarity, why do you think append should not be used here?
```
      user_dict['user_data_es'] = []
      if 'user_data' in user_dict:
          user_obj = CouchUser.wrap_correctly(user_dict)
          for key, value in user_obj.metadata.items():
              user_dict['user_data_es'].append({
                  'key': key,
                  'value': value,
              })
```
 We are actually creating a new array and appending values to it, would it matter if it is modified in place?
We are creating a new array in the dict, so append is updating the new list not the

millerdev replied:
Oh, I missed this line: `user_dict['user_data_es'] = []`

Yes, should be safe to append.

AmitPhulera commented:
@millerdev  @joxl I have created https://github.com/dimagi/commcare-hq/pull/32660 which addresses the suggestions you have made in the PR.
I am closing this one out. Would be cherry-picking some of the commits from this PR that modify the tests.
